config/cloudinit/datasource/datasource.go

```diff
@@ -30,12 +32,18 @@ type Datasource interface {
 	Finish() error
 }
 
+// TODO: move to netconf/types.go ?
+// see https://ahmetalpbalkan.com/blog/comparison-of-instance-metadata-services/
 type Metadata struct {
-	PublicIPv4    net.IP
-	PublicIPv6    net.IP
-	PrivateIPv4   net.IP
-	PrivateIPv6   net.IP
 	Hostname      string
 	SSHPublicKeys map[string]string
-	NetworkConfig interface{}
+	NetworkConfig netconf.NetworkConfig
+
+	// probably unused, but its in the initialize.env code
+	// TODO: work out if there's any reason to keep it.
+	// Lets see if anyone notices when its not set.
+	GoingPublicIPv4  net.IP
```

Comment:
So I think this is useful to keep because then the datasource can determine what $public_ipv4 is and we don't have to assume that means the first IP on the first NIC or something dumb like that.